src/games/behavmixed.imp

```diff
@@ -620,153 +544,132 @@ T MixedBehaviorProfile<T>::DiffNodeValue(const GameNode &p_node,
+  }
 
-    // push down payoffs from outcomes attached to non-terminal nodes
-    for (int child = 1; child <= node->NumChildren(); child++) {
-      m_nodeValues.SetRow(node->GetChild(child)->GetNumber(),
-			  m_nodeValues.Row(node->GetNumber()));
-    }
+  // push down payoffs from outcomes attached to non-terminal nodes
+  for (auto child: node->GetChildren()) {
+  	map_nodeValues[child] = map_nodeValues[node];
+  }
 
-    for (int pl = 1; pl <= m_support.GetGame()->NumPlayers(); pl++) {
-      m_nodeValues(node->GetNumber(), pl) = (T) 0;
-    }
+  for (auto player: m_support.GetGame()->GetPlayers()) {
+  	map_nodeValues[node][player] = T(0);
+  }
 
-    for (int child = 1; child <= node->NumChildren(); child++)  {
-      GameNode childNode = node->GetChild(child);
-      ComputeSolutionDataPass2(childNode);
+  for (auto child: node->GetChildren())  {
+  	ComputePass2_beliefs_nodeValues_actionValues(child);
 
-      GameAction act = childNode->GetPriorAction();
+  	GameAction act = child->GetPriorAction();
 
-      for (int pl = 1; pl <= m_support.GetGame()->NumPlayers(); pl++) {
-        m_nodeValues(node->GetNumber(), pl) +=
-          GetActionProb(act) * m_nodeValues(childNode->GetNumber(), pl);
-      }
+  	for (auto player: m_support.GetGame()->GetPlayers()) {
+  	map_nodeValues[node][player] += GetActionProb(act) * map_nodeValues[child][player];
+  	}
 
-      if (!iset->IsChanceInfoset()) {
-	T &cpay = m_actionValues(act->GetInfoset()->GetPlayer()->GetNumber(),
-				 act->GetInfoset()->GetNumber(),
-				 act->GetNumber());
-	if (infosetProb != infosetProb * (T) 0) {
-	  cpay += m_beliefs[node->GetNumber()] * m_nodeValues(childNode->GetNumber(), iset->GetPlayer()->GetNumber());
-	}
-	else {
-	  cpay = (T) 0;
-	}
-      }
-    }
+  	if (!iset->IsChanceInfoset()) {
+		(infosetProb != T(0)) ?
+			map_actionValues[act] += map_beliefs[node] * map_nodeValues[child][iset->GetPlayer()]:
+			map_actionValues[act] = T(0);
+  	}
   }
 }
 
-// compute realization probabilities for nodes and isets.
 template <class T>
-void MixedBehaviorProfile<T>::ComputeSolutionDataPass1(const GameNode &node) const
-{
-  if (node->GetParent()) {
-    m_realizProbs[node->GetNumber()] = m_realizProbs[node->GetParent()->GetNumber()] * GetActionProb(node->GetPriorAction());
-  }
-  else {
-    m_realizProbs[node->GetNumber()] = (T) 1;
-  }
+void MixedBehaviorProfile<T>::ComputePass3_infosetValues_regret() const
+{
+  // Populate
+  for (auto player: m_support.GetGame()->GetPlayers()) {
+    for (auto infoset : player->GetInfosets()) {
+      map_infosetValues[infoset] = T(0);
+      for (auto action : infoset->GetActions()) {
+        map_infosetValues[infoset] += GetActionProb(action) * map_actionValues[action];
+      }
 
-  if (node->GetInfoset()) {
-    for (int i = 1; i <= node->NumChildren(); i++) {
-      ComputeSolutionDataPass1(node->GetChild(i));
+      T brpayoff = map_actionValues[infoset->GetAction(1)];
```

Comment:
For future-proofing it'd be better to have

`auto actions = infoset->GetActions()`

and then initialise `brpayoff` with `actions.front()` instead of `infoset->GetAction(1)`.  Then we aren't reliant on anything involving conventions for indexing.